diff --git a/CMakeLists.txt b/CMakeLists.txt
index 47cbce7..04e4ae5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -10,6 +10,9 @@ set(CMAKE_Fortran_COMPILER ${gf_loc})
 set(CMAKE_C_COMPILER ${gcc_loc})
 set(CMAKE_CXX_COMPILER ${gpp_loc})
 
+# silence warnings temporarily 
+unset(CMAKE_COMPILE_WARNING_AS_ERROR CACHE)
+set(CMAKE_COMPILE_WARNING_AS_ERROR OFF)
 
 PROJECT(millepedeII LANGUAGES Fortran C CXX)
 set(millepedeII_VERSION 05.01.01)
