diff --git a/core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp b/core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp
index 4018e9c..53e4f2d 100644
--- a/core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp
+++ b/core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp
@@ -135,3 +135,4 @@
-            return detray::tie(m_step_aborter_state, m_momentum_aborter_state,
-                               m_pathlimit_aborter_state, m_interactor_state,
-                               m_fit_actor_state, m_updater_state);
+            return detray::tie(m_momentum_aborter_state,
+                               m_pathlimit_aborter_state, m_updater_state,
+                               m_interactor_state, m_fit_actor_state,
+                               m_step_aborter_state);
@@ -144,3 +145,4 @@
-            return detray::tie(m_step_aborter_state, m_momentum_aborter_state,
-                               m_pathlimit_aborter_state, m_fit_actor_state,
-                               m_interactor_state, m_updater_state);
+            return detray::tie(m_momentum_aborter_state,
+                               m_pathlimit_aborter_state, m_updater_state,
+                               m_fit_actor_state, m_interactor_state,
+                               m_step_aborter_state);
