diff --git a/include/alpaka/queue/sycl/QueueGenericSyclBase.hpp b/include/alpaka/queue/sycl/QueueGenericSyclBase.hpp
index 6fe66b46f568..b3eb212a87be 100644
--- a/include/alpaka/queue/sycl/QueueGenericSyclBase.hpp
+++ b/include/alpaka/queue/sycl/QueueGenericSyclBase.hpp
@@ -127,30 +127,31 @@ namespace alpaka
             }
 
             template<bool TBlocking, typename TTask>
-            auto enqueue(TTask const& task) -> void
+            auto enqueue(TTask&& task) -> void
             {
                 {
+                    using TaskType = std::decay_t<TTask>;
                     std::lock_guard<std::shared_mutex> lock{m_mutex};
 
                     clean_dependencies();
 
                     // Execute task
-                    if constexpr(is_sycl_task<TTask> && !is_sycl_kernel<TTask>) // Copy / Fill
+                    if constexpr(is_sycl_task<TaskType> && !is_sycl_kernel<TaskType>) // Copy / Fill
                     {
                         m_last_event = task(m_queue, m_dependencies); // Will call queue.{copy, fill} internally
                     }
                     else
                     {
                         m_last_event = m_queue.submit(
-                            [this, &task](sycl::handler& cgh)
+                            [this, captured_task = std::forward<TTask>(task)](sycl::handler& cgh) mutable
                             {
                                 if(!m_dependencies.empty())
                                     cgh.depends_on(m_dependencies);
 
-                                if constexpr(is_sycl_kernel<TTask>) // Kernel
-                                    task(cgh); // Will call cgh.parallel_for internally
+                                if constexpr(is_sycl_kernel<TaskType>) // Kernel
+                                    captured_task(cgh); // Will call cgh.parallel_for internally
                                 else // Host
-                                    cgh.host_task(task);
+                                    cgh.host_task(std::move(captured_task));
                             });
                     }
 
@@ -241,11 +242,11 @@ namespace alpaka
         template<concepts::Tag TTag, bool TBlocking, typename TTask>
         struct Enqueue<alpaka::detail::QueueGenericSyclBase<TTag, TBlocking>, TTask>
         {
-            static auto enqueue(alpaka::detail::QueueGenericSyclBase<TTag, TBlocking>& queue, TTask const& task)
-                -> void
+            template<typename UTask>
+            static auto enqueue(alpaka::detail::QueueGenericSyclBase<TTag, TBlocking>& queue, UTask&& task) -> void
             {
                 ALPAKA_DEBUG_MINIMAL_LOG_SCOPE;
-                queue.m_spQueueImpl->template enqueue<TBlocking>(task);
+                queue.m_spQueueImpl->template enqueue<TBlocking>(std::forward<UTask>(task));
             }
         };
 
