From f07732cc303d00416b0cc3da59b6ce1a0117e902 Mon Sep 17 00:00:00 2001
From: Paul Gessinger <paul.gessinger@cern.ch>
Date: Thu, 8 Jan 2026 09:59:32 +0100
Subject: [PATCH] Add cmath include

---
 CPU/src/TTree_hits.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/CPU/src/TTree_hits.cpp b/CPU/src/TTree_hits.cpp
index 642e8f2..d6c336c 100644
--- a/CPU/src/TTree_hits.cpp
+++ b/CPU/src/TTree_hits.cpp
@@ -11,6 +11,7 @@
  **********************************************************************/
 
 #include "TTree_hits.hpp"
+#include <cmath>
 
 
 //---------------------------------------------------------------------------------------------------------------------
-- 
GitLab

